### PR TITLE
Implement a SetBeam algorithm

### LINKED ISSUE #16228: Implement SetBeam algorithm
This is required and described by the wider absorption correction work in #15543.

The implementation here should focus on supporting a single slit-style geometry with height and width arguments specified in centimetres.

### PR DESCRIPTION
**Pull request by martyngigg**
Allows a basic rectangular slit to be defined for the beam size. Currently only used in the `MonteCarloAbsorption` algorithm. See #15543 for more details of the roadmap issue.

**To test:**

The `MonteCarloAbsorption` algorithm is the best way to test this.

``` python
from mantid.simpleapi import *

# -------------------------------------
# Comment/uncomment the required mode
#emode = 'Indirect'
emode = 'Direct'

if emode == 'Indirect':
    sample_file = 'IRS26176.RAW'
    efixed = 1.845
else:
    sample_file = 'MAR11060.RAW'
    efixed = 12.0

# Vanadium
sample_formula = 'V'
# geometry in cm
sample_height = 3.0
sample_radius = 0.1
beam_width = 0.1
beam_width = 0.1
# atoms/A^3
sample_density = 0.073

# simulation options
nlambda_cyl = 20
nlambda_mc = 20

# ------------------- Setup the workspace ------------------------

red_ws = Load(Filename=sample_file)
# Just do a single spectrum
red_ws = CropWorkspace(red_ws, StartWorkspaceIndex=2, EndWorkspaceIndex=2)
SetInstrumentParameter(red_ws, ParameterName="deltaE-mode", Value=emode, ParameterType="String")
if emode == 'Indirect':
    SetInstrumentParameter(red_ws, ParameterName="Efixed",Value=str(efixed), ParameterType="Number")
elif emode == 'Direct':
    AddSampleLog(red_ws, LogName='Ei', LogText=str(efixed), LogType='Number')
red_ws = ConvertUnits(red_ws, Target='Wavelength',
                      EMode=emode)
SetSampleMaterial(red_ws, ChemicalFormula=sample_formula,
                  SampleNumberDensity=sample_density)

# --------------------- Monte Carlo absorption cylinder ---------------------

cylinder_template = '''<cylinder id="A">
  <centre-of-bottom-base x="0.0" y="{}" z="0.0" />
  <axis x="0.0" y="1.0" z="0.0" />
  <radius val="{}" />
  <height val="{}" />
</cylinder>'''
shape_xml = cylinder_template.format(-0.5*sample_height*0.01,sample_radius*0.01, sample_height*0.01)
CreateSampleShape(red_ws, ShapeXML=shape_xml)
SetSampleMaterial(red_ws, ChemicalFormula=sample_formula,
                  SampleNumberDensity=sample_density)
SetBeam(res_ws, Geometry={'Shape': 'Slit', 'Width': beam_width, 'Height': beam_height'})
red_ws_ass_cyl_mc = MonteCarloAbsorption(InputWorkspace=red_ws,
                                     NumberOfWavelengthPoints=nlambda_mc,
                                         EventsPerPoint=10000)
```

Play with the beam size and see how it affects the results. Setting the width/height to match the sample should produce the same results as commenting out the algorithm.

Fixes #16228 

_Release notes will be updated along with the closure of the roadmap issue_

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] How do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant documentation been added/updated?
- [x] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

_Release notes will be updated along with the closure of the roadmap issue_

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the coding standards? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/test/SetBeamTest.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  //----------------------------------------------------------------------------
+  // Non-test methods
+  //----------------------------------------------------------------------------
+private:
+  Mantid::API::IAlgorithm_uptr createAlgorithm() {
+    auto alg = Mantid::Kernel::make_unique<SetBeam>();
+    alg->setChild(true);
+    alg->setRethrows(true);
+    alg->initialize();
+    return std::move(alg);
```
Comment (AntonPiccardoSelg) [inline Framework/DataHandling/test/SetBeamTest.h]: Not sure that you need to `move` here.

Comment (martyngigg) [inline Framework/DataHandling/test/SetBeamTest.h]: Won't the default be to copy as alg is an lvalue? I'm pretty sure the compiler complained when I didn't have it.

Comment (AntonPiccardoSelg) [inline Framework/DataHandling/test/SetBeamTest.h]: I think that is down to the compiler's optimization [capabilities]( stackoverflow.com/questions/17473753/c11-return-value-optimization-or-move). Not sure why it wouldn't work in this case.

Comment (martyngigg) [inline Framework/DataHandling/test/SetBeamTest.h]: True, RVO should kick in here - I'm not sure why it didn't.

**Comment by AntonPiccardoSelg** [comment]
Code Review:
 * Code looks good

Functional Testing:
 * Tested with invalid values
 * Tested with very large numbers
 * Tested with very small numbers
 * Behaves as expected, i.e. an increased beam size leads to an increased absorption

:shipit:

**Pull request merged**